### PR TITLE
Ran `pyupgrade` over codebase, added as pre-commit hook

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

### Pull request overview
- Added `pyupgrade` as a hook to pre-commit, to automatically upgrade syntactically outdated code.
- Ran `pyupgrade` over the entire codebase.

---
### The hook
The following hook was added:
```yaml
  - repo: https://github.com/asottile/pyupgrade
    rev: v2.23.3
    hooks:
    - id: pyupgrade
      args: ["--py36-plus"]
```
Note that `--py36-plus` is given as a parameter. This allows `pyupgrade` to use f-string formatting etc.

Upon committing, `pyupgrade` will run (after the default hooks, before `black`), and if it wants to modify a file, it will prevent the commit from going through, and modify the code instead. You can then `git add` again, and re-commit, now with upgraded changes.

Note that all of this only works if the user has done `pre-commit install` at some point, like is [now recommended for nltk](https://github.com/nltk/nltk/blob/develop/CONTRIBUTING.md#setting-up-a-development-environment). Only then will pre-commit do anything.

See #2775 for some discussion of adding `pyupgrade` to pre-commit, and for more discussion about pre-commit hooks.

---

### What exactly does `pyupgrade` even do?
The following is a list of the most common changes:
- Convert `%-string` formatting to `.format` or `f-string` formatting.
  - It doesn't remove all `%`-s, but does remove most. The last ones require manual work.
  - It doesn't always prefer `f-strings`, uses `.format` when that's deemed more readable or shorter.
- Convert `IOError` to `OSError`.
- Convert `str("hello")` to `"hello"`. Same with `set({...})` and `dict(...)`.
- Convert `super(ThisClass, self)....` to `super()....`.
- Convert `for element in iterable: yield element` to `yield from iterable`.

### Technical notes
- It is possible for `pre-commit` to run, fail, automatically make changes, and on the next run fail again. This can happen whenever `black` re-formats a section of code in such a way that `pyupgrade` now sees another improvement to be made. This won't be common whatsoever, but it can happen.
- This PR truly prevents ever returning to Python 3.5, as f-string formatting was only introduced in Python 3.6 as you may know.

### Other notes
- I've checked all of the changes in the first 15 changed files. I don't have the time to verify all the other 171 files, but the first 15 have given me confidence that it's all correct. 
- The tests still pass like normal.
- Performing `pre-commit run --all-files` also reformatted 8 files with `black`, and fixed trailing whitespace in 8 files. The full output (of the first run only, I ran it a few times) can be seen below:

<details>
<summary>The full output of the first pre-commit run --all-files</summary>

```
[WARNING] The 'rev' field of repo 'https://github.com/ambv/black' appears to be a mutable reference (moving tag / branch).  Mutable references are never updated after first install and are not supported.  See https://pre-commit.com/#using-the-latest-version-for-a-repository for more details.  Hint: `pre-commit autoupdate` often fixes this.
fix UTF-8 byte order marker..............................................Passed
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing nltk/tokenize/punkt.py
Fixing nltk/test/unit/files/tweets.20150430-223406.text.csv.ref
Fixing nltk/test/unit/files/tweets.20150430-223406.tweet.csv.ref
Fixing nltk/test/unit/files/tweets.20150430-223406.hashtag.csv.ref
Fixing nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref
Fixing nltk/test/unit/files/tweets.20150430-223406.user.csv.ref
Fixing .github/workflows/ci.yaml
Fixing nltk/test/unit/files/tweets.20150430-223406.retweet.csv.ref

Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing .pre-commit-config.yaml

Fix requirements.txt.....................................................Passed
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted setup.py
reformatted tools\run_doctests.py
reformatted tools\global_replace.py
reformatted tools\svnmime.py
reformatted web\conf.py
reformatted tools\nltk_term_index.py
reformatted tools\find_deprecated.py
reformatted nltk\tokenize\punkt.py
All done! \u2728 \U0001f370 \u2728
8 files reformatted, 344 files left unchanged.

pyupgrade................................................................Failed
- hook id: pyupgrade
- exit code: 1
- files were modified by this hook

Rewriting nltk/inference/mace.py
Rewriting nltk/util.py
Rewriting nltk/tokenize/__init__.py
Rewriting nltk/test/unit/translate/test_nist.py
Rewriting nltk/draw/tree.py
Rewriting nltk/tag/senna.py
Rewriting nltk/corpus/reader/util.py
Rewriting nltk/draw/util.py
Rewriting nltk/sem/cooper_storage.py
Rewriting nltk/tag/crf.py
Rewriting nltk/parse/chart.py
Rewriting nltk/translate/meteor_score.py
Rewriting nltk/cluster/em.py
Rewriting nltk/tokenize/punkt.py
Rewriting nltk/parse/evaluate.py
Rewriting nltk/tag/brill.py
Rewriting nltk/internals.py
Rewriting nltk/sem/lfg.py
Rewriting nltk/corpus/reader/chasen.py
Rewriting nltk/tokenize/nist.py
Rewriting nltk/parse/pchart.py
Rewriting nltk/sem/boxer.py
Rewriting nltk/translate/ibm5.py
Rewriting nltk/data.py
Rewriting nltk/chunk/regexp.py
Rewriting nltk/treetransforms.py
Rewriting nltk/tag/perceptron.py
Rewriting nltk/inference/nonmonotonic.py
Rewriting nltk/tokenize/texttiling.py
Rewriting nltk/corpus/reader/lin.py
Rewriting nltk/parse/shiftreduce.py
Rewriting nltk/sem/linearlogic.py
Rewriting nltk/ccg/api.py
Rewriting nltk/corpus/reader/udhr.py
Rewriting nltk/metrics/aline.py
Rewriting nltk/app/chartparser_app.py
Rewriting nltk/twitter/common.py
Rewriting setup.py
Rewriting nltk/parse/stanford.py
Rewriting nltk/corpus/reader/nombank.py
Rewriting nltk/corpus/reader/twitter.py
Rewriting nltk/sem/hole.py
Rewriting nltk/collections.py
Rewriting nltk/probability.py
Rewriting nltk/translate/ibm3.py
Rewriting nltk/tokenize/util.py
Rewriting nltk/corpus/util.py
Rewriting nltk/classify/megam.py
Rewriting nltk/tokenize/stanford.py
Rewriting nltk/tree.py
Rewriting nltk/tbl/template.py
Rewriting nltk/chat/util.py
Rewriting nltk/corpus/reader/panlex_lite.py
Rewriting nltk/corpus/reader/ipipan.py
Rewriting nltk/parse/viterbi.py
Rewriting nltk/test/unit/test_tgrep.py
Rewriting nltk/corpus/reader/switchboard.py
Rewriting nltk/test/unit/translate/test_bleu.py
Rewriting nltk/app/chunkparser_app.py
Rewriting nltk/classify/decisiontree.py
Rewriting nltk/draw/cfg.py
Rewriting nltk/tag/mapping.py
Rewriting nltk/twitter/twitterclient.py
Rewriting nltk/translate/api.py
Rewriting nltk/corpus/reader/bnc.py
Rewriting nltk/sem/chat80.py
Rewriting nltk/stem/snowball.py
Rewriting nltk/parse/featurechart.py
Rewriting nltk/cluster/util.py
Rewriting nltk/parse/dependencygraph.py
Rewriting nltk/inference/tableau.py
Rewriting nltk/app/srparser_app.py
Rewriting nltk/parse/recursivedescent.py
Rewriting nltk/chunk/named_entity.py
Rewriting nltk/corpus/reader/conll.py
Rewriting nltk/jsontags.py
Rewriting nltk/metrics/paice.py
Rewriting nltk/corpus/reader/knbc.py
Rewriting tools/nltk_term_index.py
Rewriting nltk/tokenize/regexp.py
Rewriting nltk/corpus/reader/panlex_swadesh.py
Rewriting nltk/sem/drt.py
Rewriting nltk/tokenize/simple.py
Rewriting nltk/sem/util.py
Rewriting nltk/corpus/reader/rte.py
Rewriting nltk/translate/ibm2.py
Rewriting nltk/twitter/util.py
Rewriting nltk/tbl/feature.py
Rewriting nltk/toolbox.py
Rewriting nltk/corpus/reader/childes.py
Rewriting nltk/corpus/reader/nkjp.py
Rewriting nltk/test/unit/test_corenlp.py
Rewriting nltk/featstruct.py
Rewriting nltk/tbl/rule.py
Rewriting nltk/tag/brill_trainer.py
Rewriting nltk/sentiment/util.py
Rewriting nltk/metrics/scores.py
Rewriting nltk/corpus/reader/ieer.py
Rewriting nltk/tokenize/sexpr.py
Rewriting nltk/metrics/confusionmatrix.py
Rewriting nltk/test/unit/translate/test_ibm_model.py
Rewriting nltk/tag/hmm.py
Rewriting nltk/tbl/erroranalysis.py
Rewriting nltk/test/unit/test_wordnet.py
Rewriting nltk/lm/vocabulary.py
Rewriting tools/global_replace.py
Rewriting nltk/tokenize/toktok.py
Rewriting nltk/test/unit/test_rte_classify.py
Rewriting nltk/corpus/reader/ycoe.py
Rewriting nltk/tokenize/destructive.py
Rewriting nltk/parse/malt.py
Rewriting nltk/metrics/distance.py
Rewriting tools/svnmime.py
Rewriting nltk/inference/discourse.py
Rewriting nltk/corpus/reader/timit.py
Rewriting nltk/sentiment/vader.py
Rewriting nltk/tag/sequential.py
Rewriting nltk/chunk/util.py
Rewriting nltk/parse/earleychart.py
Rewriting nltk/corpus/reader/xmldocs.py
Rewriting nltk/sentiment/sentiment_analyzer.py
Rewriting nltk/inference/resolution.py
Rewriting nltk/corpus/reader/api.py
Rewriting nltk/translate/ibm4.py
Rewriting nltk/collocations.py
Rewriting nltk/classify/scikitlearn.py
Rewriting nltk/translate/ibm1.py
Rewriting nltk/classify/util.py
Rewriting nltk/tag/stanford.py
Rewriting nltk/corpus/reader/comparative_sents.py
Rewriting nltk/tokenize/api.py
Rewriting nltk/corpus/reader/framenet.py
Rewriting nltk/test/unit/test_pos_tag.py
Rewriting nltk/corpus/reader/sentiwordnet.py
Rewriting nltk/grammar.py
Rewriting nltk/parse/generate.py
Rewriting nltk/twitter/twitter_demo.py
Rewriting nltk/tbl/demo.py
Rewriting nltk/sem/relextract.py
Rewriting nltk/tokenize/stanford_segmenter.py
Rewriting nltk/stem/regexp.py
Rewriting nltk/test/unit/test_stem.py
Rewriting nltk/sem/skolemize.py
Rewriting nltk/ccg/combinator.py
Rewriting nltk/text.py
Rewriting tools/find_deprecated.py
Rewriting nltk/test/unit/test_tokenize.py
Rewriting nltk/lm/counter.py
Rewriting nltk/__init__.py
Rewriting nltk/classify/maxent.py
Rewriting nltk/sem/glue.py
Rewriting nltk/classify/rte_classify.py
Rewriting nltk/draw/table.py
Rewriting nltk/corpus/reader/propbank.py
Rewriting nltk/corpus/reader/verbnet.py
Rewriting nltk/downloader.py
Rewriting nltk/app/wordnet_app.py
Rewriting nltk/metrics/agreement.py
Rewriting nltk/sem/evaluate.py
Rewriting nltk/inference/prover9.py
Rewriting nltk/stem/lancaster.py
Rewriting nltk/twitter/api.py
Rewriting nltk/sem/logic.py
Rewriting nltk/classify/naivebayes.py
Rewriting nltk/parse/util.py
Rewriting nltk/classify/textcat.py
Rewriting nltk/tag/util.py
Rewriting nltk/classify/weka.py
Rewriting nltk/corpus/reader/crubadan.py
Rewriting nltk/treeprettyprinter.py
Rewriting nltk/parse/projectivedependencyparser.py
Rewriting nltk/parse/corenlp.py
Rewriting nltk/corpus/reader/wordnet.py
Rewriting nltk/tokenize/treebank.py
Rewriting nltk/test/unit/lm/test_models.py
Rewriting nltk/tgrep.py
```
</details>

- Tom Aarsen

### CODE REVIEW

**Comment by stevenbird** [comment]
LGTM but I'll wait for some more careful eyes to take a look! Thanks.

**Comment by iliakur** [comment]
hmm, how does this relate to https://github.com/nltk/nltk/pull/2747?

Are you invoking a superset of the functionality used in that PR?

**Comment by tomaarsen** [comment]
@iliakur 
That PR fixed issues about [invalid escape sequences](https://github.com/asottile/pyupgrade#invalid-escape-sequences), removing unnecessary paranthesis and whitespace. 
Instead, this PR applies considerably more upgrades. It runs `pyupgrade --py36-plus`: the changes described in *What exactly does pyupgrade even do?* are now also applied.

So yes, it applies more of the functionality of pyupgrade. First and foremost, it adds pyupgrade to `pre-commit`. Then, the codebase should be updated accordingly, which is also done in this PR.

File: nltk/sem/linearlogic.py [hunk 1]
```diff
@@ -450,11 +446,7 @@ def __eq__(self, other):
         return self.d == other.d
 
     def __str__(self):
-        return (
-            "{"
-            + ", ".join("%s: %s" % (v, self.d[v]) for v in sorted(self.d.keys()))
-            + "}"
-        )
+        return "{" + ", ".join(f"{v}: {self.d[v]}" for v in sorted(self.d.keys())) + "}"
```
Comment (dannysepler) [inline nltk/sem/linearlogic.py:449]: something i learned recently, a lot of calls of dictionary `.keys()` is unnecessary. there's a funky pre-commit hook that undoes this https://github.com/mxr/unkey -- but i think we could also just call this hook on the whole nltk repo without actually adding it to pre-commit

File: nltk/tbl/erroranalysis.py [hunk 2]
```diff
@@ -32,9 +32,7 @@ def error_list(train_sents, test_sents):
             if train_pos != test_pos:
                 left = " ".join("%s/%s" % w for w in train_sent[:wordnum])
                 right = " ".join("%s/%s" % w for w in train_sent[wordnum + 1 :])
-                mid = "%s/%s->%s" % (word, test_pos, train_pos)
-                errors.append(
-                    "%25s | %s | %s" % (left[-25:], mid.center(22), right[:25])
-                )
+                mid = f"{word}/{test_pos}->{train_pos}"
```
Comment (dannysepler) [inline nltk/tbl/erroranalysis.py:35]: just remarking that `"%s/%s->%s" % (word, test_pos, train_pos)` to `f"{word}/{test_pos}->{train_pos}"` is such an improvement!

File: nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref [hunk 3]
```diff
@@ -24,10 +24,10 @@ id,text,user_mentions.id,user_mentions.screen_name
 593891106066198529,RT @GloriaDePiero: Nick Clegg is just as responsible for this Govt's failing  plan as David Cameron - he's backed the Tories all the  way,162472533,GloriaDePiero
 593891106590429185,RT @mykilmarnock: Will the person who dropped a Vote SNP badge please call at Kilmarnock police station to collect it. http://t.co/o3sG5B4L…,2787759816,mykilmarnock
 593891106787631104,RT @Markfergusonuk: The Sun’s Twitter worm appears to believe Ed Miliband is winning so far #bbcqt http://t.co/ZgZbSwnZxZ,28079313,Markfergusonuk
-593891106921799680,"@ScottishPleb @AHairyBiker 
+593891106921799680,"@ScottishPleb @AHairyBiker
```
Comment (dannysepler) [inline nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref:27]: it looks like you re-did the trailing whitespace that https://github.com/nltk/nltk/pull/2754 reverted but that's ok, https://github.com/nltk/nltk/pull/2755 should fix the error from here on out 👍

File: web/conf.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 # documentation root, use os.path.abspath to make it absolute, like shown here.
 
 # build docs using nltk from the upper dir, not the installed version
-sys.path.insert(0, os.path.abspath('..'))
```
Comment (dannysepler) [inline web/conf.py:20]: it perhaps looks like you have some black changes in this diff? these are all totally okay though!

**Comment by dannysepler** [comment]
this is really awesome to see! thank you @tomaarsen for doing all this work. i think this looks great, but i do recognize this is a big change, so i can give a day or so for other people to look at this. so excited! 🚀 🎉

File: nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref (hunk 3 above)
Comment (tomaarsen) [reply nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref:27]: You're right! I ran tests afterwards, and they passed, so I think we're in the clear!

File: web/conf.py (hunk 4 above)
Comment (tomaarsen) [reply web/conf.py:20]: Indeed. I mentioned it briefly in *Other notes*, but 8 files were affected by the whitespace trimmer, and 8 files were reformatted with `black`.

File: nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref (hunk 3 above)
Comment (purificant) [reply nltk/test/unit/files/tweets.20150430-223406.usermention.csv.ref:27]: :+1:

**Review by purificant — APPROVED** [review]

**Comment by purificant** [comment]
Great! :+1: Nice to see f-strings everywhere.

**Comment by stevenbird** [comment]
Brilliant contribution thank you @tomaarsen!
@dannysepler @iliakur @purificant thanks for your feedback

**Pull request merged**